### PR TITLE
fix(core): make trim_messages respect tool-call group integrity

### LINKED ISSUE #29637: `trim_messages` and `ChatAnthropic` token counter with tools
### Privileged issue

- [x] I am a LangChain maintainer, or was asked directly by a LangChain maintainer to create an issue here.

### Issue Content

`ChatAnthropic.get_num_tokens_from_messages` uses Anthropic's [token counting API](https://docs.anthropic.com/en/api/messages-count-tokens) and will error for invalid message sequences. These constraints are not respected inside `trim_messages`. See below for examples.

```python
from functools import partial

from langchain_anthropic import ChatAnthropic
from langchain_core.messages import AIMessage, HumanMessage, ToolMessage
from langchain_core.messages.utils import trim_messages
from langchain_core.tools import tool


@tool
def get_weather(location: str):
    """Get the weather."""
    pass

tools = [get_weather]


llm = ChatAnthropic(model="claude-3-5-sonnet-latest").bind_tools(tools)


messages = [
    HumanMessage("hi"),
    AIMessage("hello"),
    HumanMessage("what's the weather in florida?"),
    AIMessage(
        [
            {"type": "text", "text": "let's check the weahter in florida"},
            {
                "type": "tool_use",
                "id": "abc123",
                "name": "get_weather",
                "input": {"location": "florida"},
            },
        ],
        tool_calls=[
            {
                "name": "get_weather",
                "args": {"location": "florida"},
                "id": "abc123",
                "type": "tool_call",
            },
        ],
    ),
    ToolMessage(
        "It's sunny.",
        name="get_weather",
        tool_call_id="abc123",
    ),
]
```

## Example 1
```python
trim_messages(
    messages,
    max_tokens=200,
    token_counter=llm,
)
```
Breaks when we attempt to count tokens in a reversed list of messages: https://github.com/langchain-ai/langchain/blob/db8201d4dafb533133cc51c87c7ef011e546e03f/libs/core/langchain_core/messages/utils.py#L1309-L1312
> BadRequestError: Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'messages.0: `tool_result` block(s) provided when previous message does not contain any `tool_use` blocks'}}

## Example 2
```python
trim_messages(
    messages,
    max_tokens=200,
    token_counter=llm,
    strategy="first",
)
```
Breaks because `ChatAnthropic.get_num_tokens_from_messages` expects tools to be passed in as a kwarg:
https://github.com/langchain-ai/langchain/blob/db8201d4dafb533133cc51c87c7ef011e546e03f/libs/partners/anthropic/langchain_anthropic/chat_models.py#L1160-L1162
> BadRequestError: Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'Requests which include `tool_use` or `tool_result` blocks must define tools.'}}

## Example 3
```python
trim_messages(
    messages,
    max_tokens=200,
    token_counter=partial(llm.get_num_tokens_from_messages, tools=tools),
    strategy="first",
)
```
Breaks because an invalid history is sent to the token counting API:
> BadRequestError: Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'messages.3: Messages containing `tool_use` blocks must be followed by a user message with `tool_result` blocks'}}

### PR DESCRIPTION
**Pull request by giulio-leone**
## Description

`trim_messages` now treats `AIMessage(tool_calls)` + following `ToolMessage(s)` as atomic groups that cannot be split during trimming. This fixes three classes of `BadRequestError` when using providers that validate message ordering (e.g. Anthropic's token counting API).

Fixes #29637

### 1. Group-aware binary search (`_first_max_tokens`)
Binary search now operates on **group boundaries** instead of individual message indices. A group is either a single non-tool message or an `AIMessage(tool_calls)` + all following `ToolMessages`. This ensures that `messages[:mid]` never splits a tool-call exchange.

### 2. Rewritten `_last_max_tokens` (no more message reversal)
The old implementation reversed messages and delegated to `_first_max_tokens`. Reversed sequences like `[ToolMessage, AIMessage(tool_calls), ...]` are invalid for Anthropic's `count_tokens` API. The new implementation performs a **forward binary search** on group boundaries from the end, preserving natural message ordering in all calls to the token counter.

### 3. Forward bound tools to token counter
When `token_counter` is a bound LLM (via `bind_tools()`), the bound `tools` are automatically extracted from `kwargs` and forwarded to `get_num_tokens_from_messages()` — fixing the _"tool_use blocks must define tools"_ error (Example 2 from the issue).

### New helper: `_get_message_group_boundaries()`
Returns valid split-point indices where `messages[:i]` is always a structurally valid prefix — no orphaned `ToolMessages`, no `AIMessage(tool_calls)` without their replies.

## Tests
11 new tests covering:
- Tool groups never split with strategy `first` / `last`
- Token counter never receives reversed messages
- Tool groups with `include_system`, `start_on`, `end_on`
- Multiple tool groups
- Exact token boundary fit
- Bound tools forwarded to counter
- Direct `_get_message_group_boundaries()` unit test

All 165 tests pass (154 existing + 11 new).

### CODE REVIEW

**Comment by giulio-leone** [comment]
Added in 4da7f56: new test `test_trim_messages_tool_group_never_split_first_allow_partial` covering `strategy="first"` + `allow_partial=True` with a tool-call group that partially fits the budget. Verifies that even with partial trimming enabled, an AIMessage with tool_calls is never orphaned from its ToolMessages.

File: libs/core/tests/unit_tests/messages/test_utils.py [hunk 1]
```diff
@@ -2958,3 +2958,280 @@ def get_time(timezone: str) -> str:
...
+    # No orphaned tool-call messages
+    for msg in result:
+        if isinstance(msg, AIMessage) and msg.tool_calls:
+            # If an AIMessage with tool_calls is present, the following messages
+            # must include the ToolMessage(s).
+            idx = result.index(msg)
+            assert idx + 1 < len(result)
+            assert isinstance(result[idx + 1], ToolMessage)
+
+
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/core/tests/unit_tests/messages/test_utils.py:3024]: Good suggestion. The primary scenarios for tool group integrity are `strategy='last'` (where tool groups would be split by reverse-iteration) and `include_system=True` + `allow_partial=True`. `strategy='first'` with `allow_partial=True` would follow the same `_first_max_tokens` path that's already tested. I'll add a test for this combination.

**Comment by giulio-leone** [comment]
Added test for `strategy='first'` + `allow_partial=True` with a tool-call group that doesn't fit the budget (commit a74fd00). Verifies the group is excluded whole, not split.

File: libs/core/tests/unit_tests/messages/test_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply libs/core/tests/unit_tests/messages/test_utils.py:3024]: Added `test_trim_messages_tool_group_partial_boundary_first_allow_partial` in 2908dad — this test uses `max_tokens=45` which is the critical edge case where the AIMessage(tool_calls) individually fits (30+14=44 ≤ 45) but the full group does not (44+10=54 > 45). The boundary falls *inside* the tool-call group.

This also uncovered a real bug in `_first_max_tokens`: `allow_partial=True` was attempting to partially include the AIMessage with tool_calls, producing an orphaned tool-call. Fixed by adding a `_skip_partial` guard that prevents partial inclusion when `messages[idx]` is an AIMessage with `tool_calls`.

All 168 tests pass (167 existing + 1 new).

**Comment by giulio-leone** [comment]
Superseded by #35519
